### PR TITLE
Add support for streaming payload for publish packets

### PR DESCRIPTION
**Pull request by kasperl**

### CODE REVIEW

**Comment by kasperl** [comment]
Maybe you want to take a look at this too, @floitsch, at some point?

File: src/packets.toit [hunk 1]
```diff
@@ -54,6 +54,12 @@ abstract class Packet:
 
   abstract payload -> ByteArray
 
+  ensure_drained_ -> none:
+    // Most packets drain the reader they deserialize from eagerly, but
+    // for we allow streaming the payload for a few. For those, we
```
Comment (erikcorry) [inline src/packets.toit:59]:
```suggestion
    // we allow streaming the payload for a few types. For those, we
```

File: src/packets.toit [hunk 2]
```diff
@@ -252,9 +258,11 @@ class PublishPacket extends Packet:
   static TYPE ::= 3
 
   topic /string
-  payload /ByteArray
   packet_id /int?
 
+  reader_ / PublishPacketReader_? := ?
+  payload_ / ByteArray? := ?
```
Comment (erikcorry) [inline src/packets.toit:264]:
```suggestion
  reader_ /PublishPacketReader_? := ?
  payload_ /ByteArray? := ?
```

File: src/packets.toit [hunk 3]
```diff
@@ -309,6 +338,39 @@ class PublishPacket extends Packet:
         --packet_id = new_packet_id
         --duplicate = duplicate != null ? duplicate : this.duplicate
 
+class PublishPacketReader_ implements reader.SizedReader:
+  reader_ / reader.BufferedReader
+  size / int
+  remaining_ / int? := null
```
Comment (erikcorry) [inline src/packets.toit:344]:
```suggestion
  reader_ /reader.BufferedReader
  size /int
  remaining_ /int? := null
```

**Review by erikcorry — APPROVED** [review]

**Pull request merged**